eclair-core/src/main/scala/fr/acinq/eclair/blockchain/watchdogs/HeadersOverDns.scala

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2020 ACINQ SAS
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package fr.acinq.eclair.blockchain.watchdogs
+
+import akka.actor.Status
+import akka.actor.typed.scaladsl.Behaviors
+import akka.actor.typed.scaladsl.adapter.TypedActorRefOps
+import akka.actor.typed.{ActorRef, Behavior}
+import akka.io.dns.{AAAARecord, DnsProtocol}
+import akka.io.{Dns, IO}
+import fr.acinq.bitcoin.{Block, BlockHeader, ByteVector32}
+import fr.acinq.eclair.blockchain.watchdogs.BlockchainWatchdog.BlockHeaderAt
+import fr.acinq.eclair.blockchain.watchdogs.Monitoring.{Metrics, Tags}
+import org.slf4j.Logger
+import scodec.bits.BitVector
+
+/**
+ * Created by t-bast on 29/09/2020.
+ */
+
+/** This actor queries https://bitcoinheaders.net/ to fetch block headers. */
+object HeadersOverDns {
+
+  val Source = "bitcoinheaders.net"
+
+  // @formatter:off
+  sealed trait Command
+  case class CheckLatestHeaders(replyTo: ActorRef[BlockchainWatchdog.LatestHeaders]) extends Command
+  private case class WrappedDnsResolved(response: DnsProtocol.Resolved) extends Command
+  private case class WrappedDnsFailed(cause: Throwable) extends Command
+  // @formatter:on
+
+  def apply(chainHash: ByteVector32, currentBlockCount: Long): Behavior[Command] = {
+    Behaviors.setup { context =>
+      val dnsAdapters = {
+        context.messageAdapter[DnsProtocol.Resolved](WrappedDnsResolved)
+        context.messageAdapter[Status.Failure](f => WrappedDnsFailed(f.cause))
+      }.toClassic
+      Behaviors.receiveMessage {
+        case CheckLatestHeaders(replyTo) => chainHash match {
+          case Block.LivenetGenesisBlock.hash =>
+            // We try to get the next 10 blocks; if we're late by more than 10 blocks, this is bad, no need to even look further.
+            (currentBlockCount until currentBlockCount + 10).foreach(blockCount => {
+              val hostname = s"$blockCount.${blockCount / 10000}.bitcoinheaders.net"
+              IO(Dns)(context.system.classicSystem).tell(DnsProtocol.resolve(hostname, DnsProtocol.Ip(ipv4 = false, ipv6 = true)), dnsAdapters)
+            })
+            collect(replyTo, currentBlockCount, Set.empty, 10)
```

Comment:
Why don't we simply try to get the n+10th block?

Reply:
If we do that and it doesn't exist, we can't properly update the metric to know how many blocks we're missing. I think it's useful to have that metric be more precise than we're more than 10 blocks late vs we're less than 10 blocks late, even only for consistency with the other watchdogs